syntax = "proto2";

import "store/common/common-proto.proto";

package janusstore.proto;

message DependencyList {
    repeated uint64 txnid = 1;
}

message GetMessage {
    optional string key = 1;
}

message PutMessage {
    optional string key = 1;
    optional string value = 2;
}

message TransactionMessage {
    enum Status {
      PREACCEPT = 0;
      ACCEPT = 1;
      COMMIT = 2;
    }
    // TODO(eric): encode into int along with transaction.h if time
    optional string serverip = 1;
    optional uint64 serverport = 2;
    optional uint64 txnid = 3;
    optional Status status = 4;
    repeated GetMessage gets = 5;
    repeated PutMessage puts = 6;
}

message PreAcceptMessage {
    optional TransactionMessage txn = 1;
    optional uint64 ballot = 2;
}

message PreAcceptOKMessage {
    optional uint64 txnid = 1;
    optional string serverip = 2;
    optional uint64 serverport = 3;
    optional DependencyList dep = 4;
}

message PreAcceptNotOKMessage {
    optional uint64 txnid = 1;
}

message AcceptMessage {
    optional uint64 txnid = 1;
    optional DependencyList dep = 2;
    optional uint64 ballot = 3;
}

message AcceptOKMessage {
    optional uint64 txnid = 1;
}

message AcceptNotOKMessage {
    // the highest ballot number seen for the requested transaction
    optional uint64 txnid = 1;
    optional uint64 highest_ballot = 2;
}

message CommitMessage {
    optional uint64 txnid = 1;
    optional DependencyList dep = 2;
}

message CommitOKMessage {
    optional uint64 txnid = 1;
    optional bool abandon = 2;
    optional bool result = 3; // TODO may need to change this type for reads
}

message InquireMessage {
    optional uint64 txnid = 1;
}

message InquireOKMessage {
    // InquireOK sent when the inquired txn status == committing
    optional uint64 txnid = 1;
    optional DependencyList deps = 2;
}

// for failure recovery; implement if time and for now don't care
// message PrepareMessage {}
// message PrepareOKMessage {}

message Request {
     enum Operation {
          GET = 1; // TODO what is this?
          PREACCEPT = 2;
          ACCEPT = 3;
          COMMIT = 4;
          INQUIRE = 5;

          // for failure recovery
          PREPARE = 6;
     }
    optional Operation op = 1;
    optional PreAcceptMessage preaccept = 2;
    optional AcceptMessage accept = 3;
    optional CommitMessage commit = 4;
    optional InquireMessage inquire = 5;

     // for failure recovery
     // optional PrepareMessage prepare = 6;
}

message Reply {
     enum Operation {
          GET = 1; // use value
          PREACCEPT_OK = 2;
          PREACCEPT_NOT_OK = 3; // contains no message
          ACCEPT_OK = 4; // contains no message
          ACCEPT_NOT_OK = 5;
          COMMIT_OK = 6;
          INQUIRE_OK = 7;

          // for failure recovery
          PREPARE_OK = 8;
     }
    optional Operation op = 1;
    optional string value = 2; // TODO what is this for?
    optional PreAcceptOKMessage preaccept_ok = 3;
    optional PreAcceptNotOKMessage preaccept_not_ok = 4;
    optional AcceptOKMessage accept_ok = 5;
    optional AcceptNotOKMessage accept_not_ok = 6;
    optional CommitOKMessage commit_ok = 7;
    optional InquireOKMessage inquire_ok = 8;

     // for failure recovery
     // optional PrepareOKMessage prepare_ok = 7;
}
